lms/djangoapps/commerce/utils.py

```diff
@@ -109,3 +117,263 @@ def upgrade_url(self, user, course_key):
+        exceptions.Timeout: if the attempt to reach the commerce service timed out.
+    """
+    User = get_user_model()  # pylint:disable=invalid-name
+    course_key_str = unicode(course_enrollment.course_id)
+    enrollee = course_enrollment.user
+
+    service_user = User.objects.get(username=settings.ECOMMERCE_SERVICE_WORKER_USERNAME)
+    api_client = ecommerce_api_client(service_user)
+
+    log.info('Attempting to create a refund for user [%s], course [%s]...', enrollee.id, course_key_str)
+
+    refund_ids = api_client.refunds.post({'course_id': course_key_str, 'username': enrollee.username})
+
+    if refund_ids:
+        log.info('Refund successfully opened for user [%s], course [%s]: %r', enrollee.id, course_key_str, refund_ids)
+
+        _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_enrollment,
+        )
+    else:
+        log.info('No refund opened for user [%s], course [%s]', enrollee.id, course_key_str)
+
+    return refund_ids
+
+
+def _process_refund(refund_ids, api_client, course_product):
+    """
+    Helper method to process a refund for a given course_product. This method assumes that the User has already
+    been unenrolled.
+
+    Returns:
+        bool: True if the refund process was successful, False if there are any Errors that are not handled
+    """
+    config = CommerceConfiguration.current()
+
+    if config.enable_automatic_refund_approval:
+        refunds_requiring_approval = []
+
+        for refund_id in refund_ids:
+            try:
+                # NOTE: The following assumes that the user has already been unenrolled.
+                # We are then able to approve payment. Additionally, this ensures we don't tie up an
+                # additional web worker when the E-Commerce Service tries to unenroll the learner.
+                api_client.refunds(refund_id).process.put({'action': 'approve_payment_only'})
+                log.info('Refund [%d] successfully approved.', refund_id)
+            except:  # pylint: disable=bare-except
+                # Push the refund to Support to process
+                log.exception('Failed to automatically approve refund [%d]!', refund_id)
+                refunds_requiring_approval.append(refund_id)
+    else:
+        refunds_requiring_approval = refund_ids
+
+    if refunds_requiring_approval:
+        # XCOM-371: this is a temporary measure to suppress refund-related email
+        # notifications to students and support for free enrollments.  This
+        # condition should be removed when the CourseEnrollment.refundable() logic
+        # is updated to be more correct, or when we implement better handling (and
+        # notifications) in Otto for handling reversal of $0 transactions.
+        if course_product.mode != 'verified':
+            # 'verified' is the only enrollment mode that should presently
+            # result in opening a refund request.
```

Comment:
Is this comment still true? Do we not support refunding `prof-cert`/`prof-ed`/`credit`?

Reply:
Yes to my knowledge it is.